Userland/Libraries/LibWeb/Layout/FormattingContext.cpp

```diff
@@ -283,21 +283,21 @@ CSSPixelSize FormattingContext::solve_replaced_size_constraint(CSSPixels input_w
     auto& h = size.height;
 
     if (w > max_width)
-        size = { max_width, max(max_width * (h / w), min_height) };
+        size = { max_width, max(CSSPixels(max_width * (h / static_cast<double>(w))), min_height) };
```

Comment:
Seems like instead of making a fractional double here, we should reorder the operations (with parentheses for emphasis of the new order):

`size = { max_width, max((max_width * h) / w, min_height) };`

That should retain (close to?) perfect precision in this calculation. The only downside is if it's possible for the value to saturate on the high end from a high ratio.

It's worth noting that the operation order above will shift in the multiplication to put it back into its normal fixed point, and then the division will shift the fixed point back out to where it was during the multiplication operation, so we could potentially have a function to multiply by a fraction if the compiler isn't smart enough to optimize away that redundancy.

Reply:
For the real world use case I'm fixing, we overflow for your proposed solution, which is why I'm doing it this way. We currently don't have saturation implemented for multiplication either, but saturation wouldn't be a good enough behavior anyway.

Said example is https://mdn.github.io/learning-area/html/multimedia-and-embedding/mdn-splash-page-finished/index.html, `max_width` is 1200 and the image size is 1600x647.

Reply:
Wow, that size difference isn't super noticeable, good catch!

Since the overflow occurs in the integer multiplication in `operator*`, it can be solved simply by upcasting `this->raw_value()` to `i64` before the multiplication. That way, it won't truncate when multiplying, and then it will be shifted back to the normal fixed point and casted back to i32.

Just tested it locally and I get a height of 485 pixels for that image in both Firefox and Ladybird.

Reply:
Ok, I've changed it to do the `i64` cast instead. Changing associativity bothers me a bit since it's less obvious to tell why we're doing it vs. a cast to double, but on balance is probably better.

Reply:
I get you, you could certainly omit the parentheses to make it easier to grok the association between `w` and `h` too. I would think it's definitely best to avoid doing floating-point math wherever possible since fixed-point should prevent loss of precision, when we put a bit of thought into the operations and their side effects.

Reply:
Good point, it's not so bad without parentheses - done.